File: 0421-maximum-xor-of-two-numbers-in-an-array/0421-maximum-xor-of-two-numbers-in-an-array.py

class Trie:
    def __init__(self):
        self.root = {}
        self.maxi = 0
    
    def insert(self, nums):
        
        num = bin(nums)

        # Remove the '0b' prefix and keep the remaining part
        num = num[2:]

        # Ensure the string is 32 bits long by padding with zeros if needed
        while len(num) < 32:
            num = '0' + num
        
        curr =  self.root
        maxi = self.maxi
        for char in num:
            if int(char) in curr:
                curr = curr[int(char)]
            else:
                
                curr[int(char)] = {}
                curr = curr[int(char)]
                
        curr =  self.root
        store = ""
        for char in num:
            if curr:
                if (1 - int(char)) in curr:
                    store += str(1 - int(char))
                    curr =  curr[1 - int(char)]
                else:
                    store += char
                    curr =  curr[int(char)]
        compared = int(store, 2)
        self.maxi = max(self.maxi, nums ^ compared)
                

class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        trie = Trie()
        for num in nums:
            trie.insert(num)
        # print(trie.root.children)
        return trie.maxi
        



        
        

File: 0421-maximum-xor-of-two-numbers-in-an-array/0421-maximum-xor-of-two-numbers-in-an-array.cpp
#include <iostream>
#include <vector>

using namespace std;

class TrieNode {
public:
    TrieNode* children[2];

    TrieNode() {
        children[0] = nullptr;
        children[1] = nullptr;
    }
};

class Trie {
public:
    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }

    void add(int num) {
        string bits = bitset<32>(num).to_string();
        TrieNode* curr = root;

        for (char b : bits) {
            int index = b - '0';

            if (curr->children[index] == nullptr) {
                curr->children[index] = new TrieNode();
            }

            curr = curr->children[index];
        }
    }

    int maxSum(int num1) {
        string bits = bitset<32>(num1).to_string();
        TrieNode* curr = root;
        string sb = "";

        for (char b : bits) {
            int oppositeIndex = (b - '0') ^ 1; // Flipping the bit

            if (curr->children[oppositeIndex] != nullptr) {
                curr = curr->children[oppositeIndex];
                sb += to_string(oppositeIndex);
            } else {
                curr = curr->children[b - '0'];
                sb += to_string(oppositeIndex ^ 1);
            }
        }

        int num2 = stoi(sb, 0, 2);

        return num1 ^ num2;
    }
};


class Solution {
public:
    int findMaximumXOR(vector<int>& nums) {
        Trie trie;
        int maximum = 0;

        for (int num : nums) {
            trie.add(num);
            int res = trie.maxSum(num);
            maximum = max(res, maximum);
        }

        return maximum;
        
    }
};